### PR TITLE
refactor: improve error handling for Tuple.set

### LINKED ISSUE #635: Improve error message in `Tuple.set`
`Tuple.set` currently returns an error if any values given to it do not match the types that the tuple was defined to contain. Right now this error message does not mention any field names or expected/actual types, which is suboptimal, especially for large tuples where it may require a bit of work to diagnose the exact issues.

The current error is raised here: https://github.com/algorand/pyteal/blob/4c716b44819b9d529df98776423e74dac31883b1/pyteal/ast/abi/tuple.py#L354

### PR DESCRIPTION
**Pull request by shyba**
Hi, :wave: 

This is my first PR here, let me know if I missed anything.
(I saw other places where error handling could improve, but probably something for other PR/issue)

closes #635

### CODE REVIEW

File: pyteal/ast/abi/tuple.py [hunk 1]
```diff
@@ -341,17 +341,18 @@ def set(self, *values):
...
         if len(myTypes) != len(values):
             raise TealInputError(
                 f"Incorrect length for values. Expected {len(myTypes)}, got {len(values)}"
             )
-        if not all(myTypes[i] == values[i].type_spec() for i in range(len(myTypes))):
-            raise TealInputError("Input values do not match type")
+        for index, (value, myType) in enumerate(zip(values, myTypes)):
+            if not isinstance(value, BaseType):
+                raise TealInputError(f"Expected BaseType, got {value}")
+            if not myType == value.type_spec():
```
Comment (ahangsu) [inline pyteal/ast/abi/tuple.py:352]: nit: can replace with
```suggestion
            if myType != value.type_spec():
```

File: pyteal/ast/abi/tuple.py [hunk 2]
```diff
@@ -341,17 +341,18 @@ def set(self, *values):
...
+                raise TealInputError(
+                    f"Input values do not match type at {index}: {value.type_spec()} != {myType}"
```
Comment (ahangsu) [inline pyteal/ast/abi/tuple.py:354]: nit: minor err msg phrasing
```suggestion
                    f"Input values do not match type at {index=}: {value.type_spec()} != {myType}"
```

File: pyteal/ast/abi/tuple_test.py [hunk 3]
```diff
@@ -662,7 +662,7 @@ def test_Tuple_set():
     with pytest.raises(pt.TealInputError):
         tupleValue.set(uint8, uint16, uint32, uint32)
 
-    with pytest.raises(pt.TealInputError):
+    with pytest.raises(pt.TealInputError, match="at 1: uint32 != uint16"):
```
Comment (ahangsu) [inline pyteal/ast/abi/tuple_test.py:665]: if you take the change of error msg above, also take this change
```suggestion
    with pytest.raises(pt.TealInputError, match="at index=1: uint32 != uint16"):
```

**Review by ahangsu — COMMENTED** [review]
Generally lgtm, thanks for the refactoring! Just some minor suggestions.

Would be great if you could add a line in `CHANGELOG.md` (added section) in project root, about this PR.

**Comment by shyba** [comment]
Thank you for taking the time to review!
Added changelog and the proposed changes.

**Review by barnjamin — APPROVED** [review]
thanks!

**Review by ahangsu — APPROVED** [review]

**Comment by bbroder-algo** [comment]
Hi Victor, thanks for the patch!

**Pull request merged**